Release Drafter draft release tool

timkimadobe opened the pull request:
## Description
This PR adds a new GitHub action that uses the [Release Drafter](https://github.com/release-drafter/release-drafter) tool anchored at v5.x.x (latest major stable release for this tool) to create draft releases based on the latest merged PRs in the repo, using their labels to categorize into relevant groups in the changelog. 

The relevant files for this new functionality are:
.github/workflows/release-drafter-draft-release.yml -> the actual action logic itself
.github/release-drafter-draft-release-config.yml -> the configuration file that controls the draft release changelog text format

Documentation on how to use the tool and update the various configuration settings have also been added in the new `Documentation` directory:
Documentation/HowToUseReleaseDrafter.md

Currently only updating to support draft releases with these enhancements so the team can test functionality and align on the final format; the tool is also currently used for full release workflow, but the action logic and its config file have not been modified.

## Related Issue

See linked Jira ticket: [MOB-15788 - (CI/CD - Edge) Automate the release notes generation - integration with Edge proj](https://jira.corp.adobe.com/browse/MOB-15788)

## Motivation and Context

Changelog generation manually adds a lot of work to the release flow, and the current GitHub auto release generation tool in its current state is not very customizable; the Release Drafter tool helps generate well organized and customizable release changelogs to help everyone using the repo find the information they're looking for quickly and easily. 

For example, breaking changes can have their own section to be easily identifiable in the changelog at a glance. The same can be applied to new features, APIs, etc. Less impactful changes can be moved to the bottom in their own category, or excluded from the changelog altogether with a special exclusion label.

Because the tool works with the latest GitHub actions and can be incorporated in to full release workflows as another step, it is also highly portable and gives the added benefit of changelog format uniformity across repos (given that we use the same config settings for each).

## How Has This Been Tested?

This action flow has been tested using a private clone of this repo: https://github.com/timkimadobe/aepsdk-edge-ios-example

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

timkimadobe committed:
Release Drafter Draft Release tool allows for creating draft releases (and real releases if desired), using a separate config from the full repo release flow
Documentation for how to use the release drafter tool is also included

emdobrin commented on .github/release-drafter-draft-release-config.yml:
```diff
@@ -0,0 +1,34 @@
+    labels: 
+      - 'chore'
+      - 'maintenance'
+change-template: '- $TITLE @$AUTHOR (#$NUMBER)'
```
Can we use `$TITLE (#$NUMBER) @$AUTHOR` as we do for the other projects?

emdobrin commented on .github/release-drafter-draft-release-config.yml:
```diff
@@ -0,0 +1,34 @@
+      - 'fix'
+      - 'bugfix'
+      - 'bug'
+  - title: 'Maintenance'
```
Would be nice to have a section for `Deprecations`, label "deprecated"

emdobrin commented on .github/workflows/release-drafter-draft-release.yml:
```diff
@@ -0,0 +1,36 @@
+        tag: ${{ github.event.inputs.tag }}
+        version: ${{ github.event.inputs.tag }}
+        publish: ${{ github.event.inputs.create_as_draft == 'no' }}
+        config-name: release-drafter-draft-release-config.yml
```
if there are no multiple files required for this action, can we use the default: `release-drafter.yml`?

emdobrin commented on .github/release-drafter-draft-release-config.yml:
```diff
@@ -0,0 +1,34 @@
+      - 'bug'
+  - title: 'Maintenance'
+    labels: 
+      - 'chore'
```
I feel like "chore" is too generic, can we use labels like "maintenance", "documentation", "buildscripts" here?

emdobrin commented on Documentation/HowToUseReleaseDrafter.md:
````diff
@@ -0,0 +1,69 @@
+    label: 'chore'
+```
+
+Note that in the default configuration of the tool, unlabeled PRs are grouped together. The categories of PRs follow the same order as in the config yaml file, with the unlabeled category starting first, followed by the defined `categories` order.
````
Can we rather have a link to full change log and not include the uncategorized changes?

timkimadobe replied:
That sounds good, added a link to the full changelog in the config, and updated config to only include explicitly listed labels so that uncategorized PRs are not included in the changelog. Also updated the documentation to include this case with an example snippet for the config change required

emdobrin commented on Documentation/HowToUseReleaseDrafter.md:
````diff
@@ -0,0 +1,69 @@
+```yaml
+change-template: '- $TITLE @$AUTHOR (#$NUMBER)'
+```
+Given a PR whose author username is @timkimadobe and a title like:
````
Given a PR whose author is `@username` with title:

timkimadobe replied:
Updated text

emdobrin commented on Documentation/HowToUseReleaseDrafter.md:
````diff
@@ -0,0 +1,69 @@
+New API allows for deleting user data #6 
+```
+
+Results in output like:
````
The resulting output is:
```
- New API allows for deleting user data @username (#6)
```

timkimadobe replied:
Updated text, also reformatted example to use the format that aligns with how we format the changes: 
`'$TITLE (#$NUMBER) @$AUTHOR'`

emdobrin commented on Documentation/HowToUseReleaseDrafter.md:
````diff
@@ -0,0 +1,69 @@
+```
+- New API allows for deleting user data @timkimadobe (#6)
+```
+You can see that the title text is inserted into the provided format wherever the `$TITLE` variable is defined, along with the `$AUTHOR` and `$NUMBER` (PR number). The tool only provides the raw text values, so if you want to leverage GitHub's own automatic linking to users and/or PRs, include the special prefixes accordingly using the required format.
+
````
The example above seems self explanatory to me, I think we can skip this part L66-67 and just keep the link below.

timkimadobe replied:
Ok 🥲 removed the text!

emdobrin commented on Documentation/HowToUseReleaseDrafter.md:
```diff
@@ -0,0 +1,69 @@
+# GitHub release configuration
```
Thinking if we can rename this file to something like "Documentation/ReleaseProcess.md", where the Release drafter configuration is one section of it.

emdobrin commented on Documentation/HowToUseReleaseDrafter.md:
```diff
@@ -0,0 +1,69 @@
+# Changelog text and style configuration
+The Release Drafter tool uses a yaml config file to control the style of output. This config controls changelog text level elements; to control GitHub release level elements (tags, release name, draft/full release status, etc.), see [GitHub release configuration](#github-release-configuration).
+
+By default, the Release Drafter action looks at the repository's ~/.github/ directory (NOT the same level as the workflow itself; one level above) for a file with the name `release-drafter.yml`. If a different filename is desired, an argurment can be added to the `with` parameters (other `with` parameters not included in example snippet for brevity):  
```
Looking to see if we can simplify this one:

By default, the Release drafter action looks for its configuration at `.github/release-drafter.yml`. To use a different config filename, specify the `config-name` as in the example below:

timkimadobe replied:
Thanks, I think that your suggested wording is much more concise. I did want to emphasize that the config is in a separate directory from the workflow file itself though; what do you think about:

By default, the Release Drafter action looks for its configuration at `.github/release-drafter.yml` (note this is one level up from the action's own `workflows` directory). To use a different config filename, specify the `config-name` as in the example below:

emdobrin reviewed (commented):
Nice work @timkimadobe! 🎉  
Have you also looked into including this into the existing release workflow?

timkimadobe committed:
Update config to only include explicitly labeled PRs, full changelog link at the bottom
Update documentation to reflect feedback and updated output style
Remove draft-specific config, and draft tool points to actual changelog config (same as prd)

timkimadobe reviewed (commented):
Thanks for the speedy feedback @emdobrin! I've made the updates based on your feedback, but had one question about a documentation change https://github.com/adobe/aepsdk-edgeidentity-ios/pull/64/files#r856525169 

With the config changes now going to the actual release-drafter.yml file, when this PR is merged into a branch that uses the full release flow, the Release Drafter in that flow will reference this config

emdobrin commented on .github/workflows/release-drafter-draft-release.yml:
```diff
@@ -0,0 +1,36 @@
+name: Create Draft Release (GitHub only)
```
nit: File rename  release-drafter-draft-release -> draft-release.yml

timkimadobe replied:
Updated filename!

emdobrin commented on Documentation/ReleaseProcess.md:
````diff
@@ -0,0 +1,83 @@
+    label: 'chore'
+```
+
+Note that in the default configuration of the tool, unlabeled PRs are grouped together. The categories of PRs follow the same order as in the config yaml file, with the unlabeled category starting first, followed by the defined `categories` order. If unlabeled PRs are not desired in the changelog, explicitly define the full set of labels that should be included by using the `include-labels` prope ...
````
Rewording suggestion: "If unlabeled PRs are not desired in the changelog, explicitly define the full set of labels that should be included by using the `include-labels` property at the top level of the config file:"
->
If you want to exclude unlabeled PRs, define the `include-labels` option as below:

timkimadobe replied:
Updated according to suggestion!

emdobrin commented on Documentation/ReleaseProcess.md:
```diff
@@ -0,0 +1,83 @@
+# GitHub release configuration
```
Under this can you include a new subtitle h2 "## Generate automatic release notes using Release Drafter" (and then also move the other subtitles one level down)?
This will allow us to update this document in the future with other details regarding our release process.

timkimadobe replied:
That makes sense, I've updated to include the header you suggested and shifted the children headers accordingly

emdobrin reviewed (approved):
Left few renaming/rewording suggestions, but other than that it looks great

timkimadobe committed:
Update documentation for release drafter: header levels and some text updates
Update release drafter config to create Enhancements category and remove redundant bug-type labels

timkimadobe reviewed (commented):
Thanks for the followup review @emdobrin! I've updated the documentation using your feedback, and also updated the config based on the feedback from the demo:
- removed redundant bug-type labels
- created a new Enhancements category using the `enhancement` label

timkimadobe committed:
Merge branch 'dev' into github-actions-release-drafter-update
